src/coding_theory/hamming.lean

```diff
@@ -0,0 +1,308 @@
+
+/-- Corresponds to `norm_eq_zero`. -/
+@[simp] lemma hamm_norm_eq_zero_iff {x : Π i, β i} : hamm_norm x = 0 ↔ x = 0 :=
+hamm_dist_eq_zero
+
+/-- Corresponds to `norm_ne_zero_iff`. -/
+lemma hamm_norm_ne_zero_iff {x : Π i, β i} : hamm_norm x ≠ 0 ↔ x ≠ 0 :=
+hamm_dist_ne_zero
+
+/-- Corresponds to `norm_pos_iff`. -/
+lemma hamm_norm_pos_iff {x : Π i, β i} : 0 < hamm_norm x ↔ x ≠ 0 := hamm_dist_pos
+
+@[simp] lemma hamm_norm_lt_one {x : Π i, β i} : hamm_norm x < 1 ↔ x = 0 := hamm_dist_lt_one
+
+lemma hamm_norm_le_card_fintype {x : Π i, β i} : hamm_norm x ≤ fintype.card ι :=
+hamm_dist_le_card_fintype
+
+lemma hamm_norm_comp_le_hamm_norm (f : Π i, γ i → β i) {x : Π i, γ i} (hf : Π i, f i 0 = 0) :
+  hamm_norm (λ i, f i (x i)) ≤ hamm_norm x :=
+begin
+  refine eq.trans_le _ (hamm_dist_comp_le_hamm_dist f),
+  simp_rw [hamm_norm, pi.zero_def, hf],
+end
+
+lemma hamm_norm_comp (f : Π i, γ i → β i) {x : Π i, γ i} (hf₁ : Π i, injective (f i))
+  (hf₂ : Π i, f i 0 = 0) : hamm_norm (λ i, f i (x i)) = hamm_norm x :=
+begin
+  simp_rw ← hamm_dist_zero_right, convert hamm_dist_comp f hf₁,
+  simp_rw [pi.zero_apply, hf₂], refl
+end
+
+lemma hamm_norm_smul_le_hamm_norm [has_zero α] [Π i, smul_with_zero α (β i)] {k : α}
+  {x : Π i, β i} : hamm_norm (k • x) ≤ hamm_norm x :=
+hamm_norm_comp_le_hamm_norm (λ i (c : β i), k • c) (λ i, by simp_rw smul_zero')
+
+lemma hamm_norm_smul [has_zero α] [Π i, smul_with_zero α (β i)] {k : α}
+  (hk : ∀ i, is_smul_regular (β i) k) (x : Π i, β i) : hamm_norm (k • x) = hamm_norm x :=
+hamm_norm_comp (λ i (c : β i), k • c) hk (λ i, by simp_rw smul_zero')
+
+end has_zero
+
+lemma hamm_dist_eq_hamm_norm [Π i, add_group (β i)] (x y : Π i, β i) :
+  hamm_dist x y = hamm_norm (x - y) :=
+by simp_rw [  ← hamm_dist_zero_right, hamm_dist, pi.sub_apply,
+              pi.zero_apply, sub_ne_zero]
+
+end hamm_dist_wt
+
+/-! ### The `hamm` type synonym -/
+
+/--
+Type synonym for a Pi type which we equip with the Hamming metric, adding relevant
+instances as needed.
+-/
+def hamm {ι : Type*} (β : ι → Type*) : Type* := Π i, β i
+
+namespace hamm
+
+variables {α ι : Type*} {β : ι → Type*}
+
+/- Defines instances that do not require the distance function. -/
+
+instance [Π i, inhabited (β i)] : inhabited (hamm β) := ⟨λ i, default⟩
+instance [decidable_eq ι] [fintype ι] [Π i, fintype (β i)] : fintype (hamm β) := pi.fintype
+instance [inhabited ι] [inst : ∀ i, nonempty (β i)] [nontrivial (β default)] : nontrivial (hamm β)
+:= pi.nontrivial
+instance [fintype ι] [Π i, decidable_eq (β i)] : decidable_eq (hamm β) :=
+fintype.decidable_pi_fintype
+
+instance [Π i, has_zero (β i)] : has_zero (hamm β) := pi.has_zero
+instance [Π i, has_one (β i)] : has_one (hamm β) := pi.has_one
+instance [Π i, has_neg (β i)] : has_neg (hamm β) := pi.has_neg
+instance [Π i, has_add (β i)] : has_add (hamm β) := pi.has_add
+instance [Π i, has_sub (β i)] : has_sub (hamm β) := pi.has_sub
+instance [Π i, has_mul (β i)] : has_mul (hamm β) := pi.has_mul
```

Comment:
I'd probably drop `has_one` and `has_mul`, since you haven't added `monoid`, `mul_zero_class`, `mul_one_class`, `comm_monoid`, `comm_semigroup`, `monoid_with_zero`, `mul_zero_one_class`, `semiring`, `comm_semiring`, ...

Clearly you have an application in mind for addition of hamming codes, but given the substantial amount of boilerplate involved, I wouldn't add the support for multiplication unless you need it already

Reply:
My thinking was that certainly in some cases (over F_2 is what I was thinking) there is a sensible such notion (and multiplication corresponds to bitwise and). But sure, I'll remove it for now.